lib/analytics/page.rb

```diff
@@ -0,0 +1,72 @@
+module Analytics
+  class Page
+    def self.increment(page_id, new_member:)
+      new(page_id).increment_actions(new_member: new_member)
+    end
+
+    def initialize(page_id)
+      @page_id = page_id
+    end
+
+    def total_actions(new_members: false)
+      Analytics.store.get( key(new_members )).to_i
+    end
+
+    def total_actions_over_time(period: :hour, new_members: false)
+      send("total_by_#{period}", new_members)
+    end
+
+    def increment_actions(new_member: false)
+      if new_member
+        incr( key(new_member) )
+        incr( key_with_hour( new_member: true ) )
+        incr( key_with_day(  new_member: true ) )
+      end
+
+      incr(key)
+      incr( key_with_hour )
+      incr( key_with_day  )
+    end
+
+    private
+
+    def total_by_hour(new_members_only)
+      12.times.inject({}) do |memo, i|
+        hour = (Time.now.utc - i.send(:hour)).beginning_of_hour.to_s(:db)
+
+        memo[hour] = Analytics.store.get(key_with_hour(hour: hour, new_member: new_members_only)).to_i
+        memo
+      end
+    end
+
+    def total_by_day(new_members_only)
+      30.times.inject({}) do |memo, i|
+        date = Time.now.utc - i.send(:day)
+
+        day = (Time.now.utc - i.send(:day)).beginning_of_day.to_s(:db)
+
+        memo[day] = Analytics.store.get(key_with_day(day: date.day, new_member: new_members_only)).to_i
+        memo
+      end
+    end
+
+    def key(new_member = false)
```

Comment:
Any redis data type takes a string for its address. 

Calling `key()` would return `"pages:1:total_actions"` - which is where I keep count of all actions for a page. `key(new_member: true)` returns `"pages:1:total_actions:new_members"` which is where I keep count of actions from new members only.

```ruby
    def key(new_member = false)
      "pages:#{@page_id}:total_actions".tap do |key|
        key << ":new_members" if new_member
      end

      # without tap:
      key = "pages:#{@page_id}:total_actions"
      key << ":new_members" if new_member
  end
```